File: benchmarks/fixtures/webapp_ts/src/api/v1/payments.ts
/** API v1 payment endpoints. */
import { getLogger, validateRequest } from '../../utils/helpers';
import { validate as validatePayment } from '../../validators/payment';
import { PaymentProcessor } from '../../services/payment/processor';
import { DatabaseConnection } from '../../database/connection';
import { EventDispatcher } from '../../events/dispatcher';

const logger = getLogger("api.v1.payments");

/** Create payment. */
export async function handleCreatePayment(request: Record<string, unknown>, db: DatabaseConnection, events: EventDispatcher): Promise<Record<string, unknown>> {
    validateRequest(request);
    const body = request["body"] as Record<string, unknown>;
    const paymentData = validatePayment(body);
    const processor = new PaymentProcessor(db, events);
    processor.initialize();
    const result = await processor.processPayment(
        paymentData["user_id"] as string,
        paymentData["amount"] as number,
        paymentData["currency"] as string,
    );
    return { status: 201, data: result };
}

/** Refund payment. */
export async function handleRefund(request: Record<string, unknown>, db: DatabaseConnection, events: EventDispatcher): Promise<Record<string, unknown>> {
    validateRequest(request);
    const body = request["body"] as Record<string, unknown>;
    const processor = new PaymentProcessor(db, events);
    processor.initialize();
    const result = await processor.refund(body["transaction_id"] as string, (body["reason"] as string) ?? "");
    return { status: 200, data: result };
}

File: benchmarks/fixtures/webapp_py/api/v2/payments.py
"""API v2 payment endpoints — adds webhook support."""

from typing import Any, Dict, List

from ...utils.logging import get_logger
from ...utils.helpers import validate_request
from ...validators.payment import validate as validate_payment_data
from ...services.payment.processor import PaymentProcessor
from ...database.connection import DatabaseConnection
from ...events.dispatcher import EventDispatcher
from ...exceptions import PaymentError, ValidationError

_logger = get_logger("api.v2.payments")


def handle_create_payment(request: Dict[str, Any], db: DatabaseConnection,
                          events: EventDispatcher) -> Dict[str, Any]:
    """Handle v2 payment creation with idempotency key."""
    validate_request(request)
    body = request.get("body", {})
    idempotency_key = request.get("headers", {}).get("Idempotency-Key", "")

    _logger.info(f"API v2 create payment (idempotency={idempotency_key[:12]}...)")

    payment_data = validate_payment_data(body)

    processor = PaymentProcessor(db, events)
    processor.initialize()

    result = processor.process_payment(
        user_id=payment_data["user_id"],
        amount=payment_data["amount"],
        currency=payment_data["currency"],
        payment_method=payment_data.get("payment_method", "card"),
    )

    return {"status": 201, "data": result}


def handle_webhook(request: Dict[str, Any], db: DatabaseConnection,
                  events: EventDispatcher) -> Dict[str, Any]:
    """Handle payment gateway webhook callbacks."""
    validate_request(request)
    body = request.get("body", {})

    event_type = body.get("type", "")
    _logger.info(f"Payment webhook: {event_type}")

    processor = PaymentProcessor(db, events)
    processor.initialize()

    if event_type == "payment.succeeded":
        txn_id = body.get("data", {}).get("transaction_id", "")
        # Already processed — just acknowledge
        return {"status": 200, "data": {"acknowledged": True}}
    elif event_type == "payment.failed":
        _logger.info(f"Payment failed webhook: {body}")
        return {"status": 200, "data": {"acknowledged": True}}
    else:
        _logger.info(f"Unknown webhook event: {event_type}")
        return {"status": 200, "data": {"acknowledged": True}}


def handle_revenue_report(request: Dict[str, Any], db: DatabaseConnection,
                          events: EventDispatcher) -> Dict[str, Any]:
    """Generate revenue report — v2 only."""
    validate_request(request)
    params = request.get("params", {})

    start = params.get("start_date", "2024-01-01")
    end = params.get("end_date", "2024-12-31")

    processor = PaymentProcessor(db, events)
    processor.initialize()

    report = processor.revenue_report(start, end)
    return {"status": 200, "data": report}
